src/core/auth/qgsauthconfigurationstorage.h

```diff
@@ -0,0 +1,462 @@
+     * \param id sha hash
+     * \param hostport string host:port
+     * \return a SSL certificate custom config
+     */
+    virtual const QgsAuthConfigSslServer loadSslCertCustomConfig( const QString &id, const QString &hostport ) const = 0;
+
+    /**
+     * Loads an SSL certificate custom config by \a hostport (host:port)
+     * \param hostport host:port
+     * \return a SSL certificate custom config
+     */
+    virtual const QgsAuthConfigSslServer loadSslCertCustomConfigByHost( const QString &hostport ) const = 0;
+
+    /**
+     * \brief sslCertCustomConfigs get SSL certificate custom configs
+     * \return list of SSL certificate custom config
+     */
+    virtual const QList<QgsAuthConfigSslServer> sslCertCustomConfigs() const = 0;
+
+    /**
+     * Returns the list of SSL certificate custom config ids.
+     * \return list of SSL certificate custom config ids
+     */
+    virtual QStringList sslCertCustomConfigIds() const = 0;
+
+    //! Check if SSL certificate custom config exists
+    virtual bool sslCertCustomConfigExists( const QString &id, const QString &hostport ) = 0;
+
+    //! Remove an SSL certificate custom config
+    virtual bool removeSslCertCustomConfig( const QString &id, const QString &hostport ) = 0;
+
+    //! Store a certificate authority
+    virtual bool storeCertAuthority( const QSslCertificate &cert ) = 0;
+
+    /**
+     * \brief certAuthority get a certificate authority by \a id (sha hash)
+     * \param id sha hash
+     * \return a certificate
+     */
+    virtual const QSslCertificate loadCertAuthority( const QString &id ) const = 0;
+
+    //! Check if a certificate authority exists
+    virtual bool certAuthorityExists( const QSslCertificate &cert ) const = 0;
+
+    //! Remove a certificate authority
+    virtual bool removeCertAuthority( const QSslCertificate &cert ) = 0;
+
+    //! Returns the map of CA certificates hashes in the storages and their trust policy.
+    virtual const QMap<QString, QgsAuthCertUtils::CertTrustPolicy> caCertsPolicy() const = 0;
+
+    //! Returns the list of CA certificates in the storage
+    virtual const  QList<QSslCertificate> caCerts() const = 0;
+
+    //! Store certificate trust policy
+    virtual bool storeCertTrustPolicy( const QSslCertificate &cert, QgsAuthCertUtils::CertTrustPolicy policy ) = 0;
+
+    //! Load certificate trust policy
+    virtual QgsAuthCertUtils::CertTrustPolicy loadCertTrustPolicy( const QSslCertificate &cert ) const = 0;
+
+    //! Remove certificate trust policy
+    virtual bool removeCertTrustPolicy( const QSslCertificate &cert ) = 0;
+
+    //! Check if certificate trust policy exists
+    virtual bool certTrustPolicyExists( const QSslCertificate &cert ) const = 0;
+
+#endif
+
+    /**
+     * Returns the list of (encrypted) master passwords stored in the database.
+     * \returns list of master passwords
+     */
+    virtual const QList<QgsAuthConfigurationStorage::MasterPasswordConfig> masterPasswords( ) const = 0;
```

Comment:
Yes, good point.

Reply:
They can't be skipped because they are pure virtual and sip fails.
Besides, the passwords are stored encrypted (only the hash is stored) and they are available by a direct read of the DB anyway.

Reply:
> They can't be skipped because they are pure virtual and sip fails.


Just add "SIP_ABSTRACT" to the class definition -- eg QgsClassificationMethod .

I'd still avoid exposing any potentially sensitive method like this until there's a specific justified reason to expose.

Reply:
My original implementation was with SIP_ABSTRACT but when I use it I cannot create a subclass in python and it fails with:
TypeError: qgis._core.QgsAuthConfigurationStorage cannot be instantiated or sub-classed
Is there any workaround besides removing all pure virtual methods from QgsAuthConfigurationStorage ?